students/alheadstrong/lesson_4/trigram.py

```diff
@@ -0,0 +1,99 @@
+import random
+from tkinter import *
+
+
+def generate_dictionary(input_text):
+    """Return a dictionary with keys consisting of each two word combination in the text as a tuple. Values are lists
+    of each one word string that follows the two word key.
+    """
+    d = {}
+    delimiters = '\n', ' ', '\t', '\\r\\n', '\r', '\\xe2\\x80\\x9c', '\\xe2\\x80\\x9d'
+    form_string = '|'.join(map(re.escape, delimiters))
+    t_list = re.split(form_string, input_text)  # Using 're' module for multiple string delimiters.
+
+    for word in t_list[:]:
+        if word == '':
+            t_list.remove(word)  # Remove empty string objects from list.
+
+    for i in range(len(t_list)-2):
+        key = tuple(t_list[i:i+2])
+        if key in d.keys():
+            d[key] = d[key]+[t_list[i+2]]
+        else:
+            d[key] = [t_list[i+2]]
+    return d
+
+
+def generate_start_key(d):
+    """Return a two-word key that begins a sentence from a dictionary."""
+    start_words = []
+    for key in d.keys():
+        if key[0][-1:] == ('.' or '?' or '!') and len(key[1:]) > 0:
```

Comment:
are you check here the the value returned by `key[0][-1:]` is either `'.'`  or `'?'` or `'!'` ?
that's not what this is doing, instead you want:
```
if key[0][-1:] in ('.' or '?' or '!')
```

I can explain further if you'd like me to.

Reply:
yes, it seems like it could work but it does something a little different.

in case of `a or b or c` it will test each var at a time, and will stop at first "true" var .. so in your case `'.'` will always evaluate to true so it will only ever compare against `'.'`